src/player.ts

```diff
@@ -0,0 +1,57 @@
+import { pull, omit } from 'lodash';
+
+type RankTable = { [playername: string]: number };
+
+export type Player = {
+    name: string;
+    candidates: Player[];
+    rankTable: RankTable;
+    capacity: number;
+};
+
+export const createPlayer = (name: string, capacity = 1): Player => ({
+    name,
+    candidates: [], // list of candidates
+    rankTable: {}, // associate a preference rank to each candidate's name
+    capacity,
+});
+
+export const addCandidate = (
+    { name, candidates, rankTable, capacity }: Player,
+    candidate: Player,
+): Player => ({
+    name,
+    candidates: [...candidates, candidate],
+    rankTable: { ...rankTable, [candidate.name]: candidates.length },
+    capacity,
+});
+
+export const addCandidates = (player: Player, candidates: Player[]): Player =>
+    candidates.reduce((acc, candidate) => addCandidate(acc, candidate), player);
+
+export const removeCandidate = (
+    { name, candidates, rankTable, capacity }: Player,
+    candidate: Player,
+): Player => ({
+    name,
+    candidates: pull(candidates, candidate),
```

Comment:
Be careful PEM, the pull function mutates its arguments, prefer the usage of `without`: https://lodash.com/docs/4.17.15#without